tensorflow/python/layers/base.py

```diff
@@ -450,9 +453,61 @@ def _should_add_regularizer(variable, existing_variable_set):
             **kwargs)
 
         if regularizer:
-          if (ops.executing_eagerly_outside_functions()
-              or _should_add_regularizer(variable, existing_variables)):
-            self._handle_weight_regularization(name, variable, regularizer)
+
+          # In order to support DistributionStrategy, we move regularization
+          # in the following `add_regularizer` merge_fn. Regularization should
+          # be declared according to variable type. For example, we should call
+          # regularizer on all repilcas when it is a `MirroredVariable`.
+          # However, `MirroredVariable` is created when first replica thread is
+          # called, which makes regularizer statements on other replica
+          # thread ignored because of the `existing_variabale` check. Therefore,
+          # it is reasonable to put regularization statement in
+          # `cross_replica` context.
+          def add_regularizer(_, should_add_regularizer, variable, regularizer,
+                              name):
+            if isinstance(should_add_regularizer, value_lib.PerReplica):
+              should_add_regularizer = any(should_add_regularizer.values)
+            if should_add_regularizer:
+              if not isinstance(variable, value_lib.DistributedValues):
+                regularizer = regularizer.primary if \
+                    isinstance(regularizer, value_lib.DistributedValues) else \
+                    regularizer
+                self._handle_weight_regularization(name, variable, regularizer)
+                return
+
+              g = ops.get_default_graph()
+              l_before = g.get_collection(ops.GraphKeys.REGULARIZATION_LOSSES)
+              for d in variable.devices:
+                self._handle_weight_regularization(
+                    name, variable.get(d), regularizer.get(d))
+
+              # Since the regularization loss operations have been added to
+              # global collection, so we manually remove them and replace with
+              # the `Mirrored` type.
```

Comment:
I am not sure I understand why this needs to be done. can you give me an example of your use case for adding this?

Reply:
Like `MirroredVariable`, the `regularizer` should be formated in `DistributedValues` not the atomic type.  So If I get values from  `tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSSES)`, I can get its value on current device.

Reply:
Like MirroredVariable, the regularizer should be formated in DistributedValues not the atomic type. So If I get values from tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSSES), I can get its value on current device.